src/driver-openinput.c

```diff
@@ -0,0 +1,483 @@
+
+#define OI_REPORT_MAX_SIZE		OI_REPORT_LONG_SIZE
+#define OI_REPORT_DATA_INDEX		3
+#define OI_REPORT_DATA_MAX_SIZE		OI_REPORT_LONG_SIZE - OI_REPORT_DATA_INDEX
+
+/* protocol function pages */
+#define OI_PAGE_INFO			0x00
+#define OI_PAGE_GIMMICKS		0xFD
+#define OI_PAGE_DEBUG			0xFE
+#define OI_PAGE_ERROR			0xFF
+
+/* info page (0x00) functions */
+#define OI_FUNCTION_VERSION			0x00
+#define OI_FUNCTION_FW_INFO			0x01
+#define OI_FUNCTION_SUPPORTED_FUNCTION_PAGES	0x02
+#define OI_FUNCTION_SUPPORTED_FUNCTIONS		0x03
+
+/* error page (0xFF) */
+#define OI_ERROR_INVALID_VALUE		0x01
+#define OI_ERROR_UNSUPPORTED_FUNCTION	0x02
+#define OI_ERROR_CUSTOM			0xFE
+
+
+static unsigned int report_rates[] = { 125, 250, 500, 750, 1000 };
+
+struct openinput_drv_data {
+	unsigned int num_profiles;
+	unsigned int num_resolutions;
+	unsigned int num_buttons;
+	unsigned int num_leds;
+	unsigned int fw_major;
+	unsigned int fw_minor;
+	unsigned int fw_patch;
+	uint64_t supported;
+};
+
+struct oi_report_t {
+	uint8_t id;
+	uint8_t function_page;
+	uint8_t function;
+	uint8_t data[29];
+} __attribute__((__packed__));
+
+
+/* gcc is being picky and asking for the prototypes to bve defined (?) */
+const char* openinput_function_page_get_name(uint8_t page);
+const char* openinput_function_get_name(uint8_t page, uint8_t function);
+
+#define CASE_RETURN_STRING(a) case a: return #a; break
+const char* openinput_function_page_get_name(uint8_t page)
+{
+	static char numeric[16];
+	char *str;
+
+	switch(page)
+	{
+	CASE_RETURN_STRING(OI_PAGE_INFO);
+	CASE_RETURN_STRING(OI_PAGE_GIMMICKS);
+	CASE_RETURN_STRING(OI_PAGE_DEBUG);
+	CASE_RETURN_STRING(OI_PAGE_ERROR);
+	default:
+		sprintf_safe(numeric, "0x%02x", page);
+		str = numeric;
+		break;
+	}
+
+	return str;
+}
+
+const char* openinput_function_get_name(uint8_t page, uint8_t function)
+{
+	static char numeric[32];
+	char *str = NULL;
+
+	switch(page)
+	{
+	case OI_PAGE_INFO:
+		switch (function)
+		{
+		CASE_RETURN_STRING(OI_FUNCTION_VERSION);
+		CASE_RETURN_STRING(OI_FUNCTION_FW_INFO);
+		CASE_RETURN_STRING(OI_FUNCTION_SUPPORTED_FUNCTION_PAGES);
+		CASE_RETURN_STRING(OI_FUNCTION_SUPPORTED_FUNCTIONS);
+		}
+	}
+
+	if (!str)
+	{
+		sprintf_safe(numeric, "0x%02x 0x%02x", page, function);
+		str = numeric;
+	}
+
+	return str;
+}
+#undef CASE_RETURN_STRING
+
+static const char* openinput_get_error_string(struct oi_report_t *report)
+{
+	char *str = "", help_str[OI_REPORT_LONG_SIZE - OI_REPORT_DATA_INDEX + 1];
+
+	switch (report->function)
+	{
+		case OI_ERROR_INVALID_VALUE:
+			asprintf(&str, "Invalid value (in position %u)", report->data[2]);
+			break;
+		case OI_ERROR_CUSTOM:
+			memcpy(help_str, report->data, sizeof(help_str)); /* copy error string */
+			help_str[sizeof(help_str)-1] = '\0'; /* terminate string */
+			asprintf(&str, "Custom error (%s)", help_str);
+			break;
+		case OI_ERROR_UNSUPPORTED_FUNCTION:
+			asprintf(&str, "Unsupported function (0x%02x, 0x%02x)", report->data[0], report->data[1]);
```

Comment:
printf directive `%#x` prints the `0x` prefix, so `%02#x` in this case, iirc.

Reply:
When using this in `xasprintf`, it gives me:
```
../src/driver-openinput.c:139:45: warning: unknown conversion type character ‘#’ in format [-Wformat=]
  139 |   xasprintf(&str, "Unsupported function (%02#x, %02#x)", report->data[0], report->data[1]);
      |                                             ^
../src/driver-openinput.c:139:52: warning: unknown conversion type character ‘#’ in format [-Wformat=]
  139 |   xasprintf(&str, "Unsupported function (%02#x, %02#x)", report->data[0], report->data[1]);
      |                                                    ^
../src/driver-openinput.c:139:19: warning: too many arguments for format [-Wformat-extra-args]
  139 |   xasprintf(&str, "Unsupported function (%02#x, %02#x)", report->data[0], report->data[1]);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The `sprintf_safe` usages seem to be fine.

Reply:
Well, they are not fine, they are printing the literal. Seems like `%#02x` works there. But I noticed that using `#` will not add the prefix when the value is 0 and will not correctly pad with 0 like it previously did.

```
ratbag debug: openinput: found function 00 00
ratbag debug: openinput: found function 00 0x1
ratbag debug: openinput: found function 00 0x2
ratbag debug: openinput: found function 00 0x3
```

Going back to the old format:

```
ratbag debug: openinput: found function 0x00 0x00
ratbag debug: openinput: found function 0x00 0x01
ratbag debug: openinput: found function 0x00 0x02
ratbag debug: openinput: found function 0x00 0x03
```

So I think I'll stick with that :sweat_smile:

Reply:
> But I noticed that using # will not add the prefix when the value is 0 

urgh, true. forgot about that. fun fact, `%#X` will uppercase hex but it will also uppercase the `X` to `0X12AB` which is probably the most useless feature ever...

> and will not correctly pad with 0 like it previously did.

the `0x` counts towards the length so we would need `%#04x` here to get `0x01`. But yeah, the zero is the killer here.